### PR TITLE
Add CUDA support for extruded spaces

### PR DESCRIPTION
**Pull request by charleskawczynski**
Closes https://github.com/CliMA/ClimaCore.jl/issues/1157.

Co-authored with @simonbyrne.

Supersedes #1168, #1176, #1174.

- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by charleskawczynski** [comment]
bors try

File: test/Spaces/extruded_cuda.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    # Test that all geometries match with CPU version:
+    @test_broken all(
+        parent(cpuspace.center_local_geometry) .==
+        Array(parent(gpuspace.center_local_geometry)),
+    )
+    @test_broken all(
+        parent(cpuspace.face_local_geometry) .==
+        Array(parent(gpuspace.face_local_geometry)),
+    )
```
Comment (charleskawczynski) [inline test/Spaces/extruded_cuda.jl:47]: Well, I'm glad we added these tests, @simonbyrne, these tests are failing.

**Comment by simonbyrne** [comment]
Something isn't right with broadcasting:
```
julia> parent(gpuspace.center_local_geometry.coordinates.z)[:, 1, 1, :, 1]
10×1 CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}:
 0.05
 0.05
 0.05
 0.05
 0.05
 0.05
 0.05
 0.05
 0.05
 0.05

julia> parent(cpuspace.center_local_geometry.coordinates.z)[:, 1, 1, :, 1]
10×1 Matrix{Float64}:
 0.05
 0.15000000000000002
 0.25
 0.35
 0.45
 0.55
 0.6499999999999999
 0.75
 0.8500000000000001
 0.95
```

**Comment by charleskawczynski** [comment]
Alright, @simonbyrne found that we were basically never calling `getindex` on slabs, which was implemented incorrectly. It's fixed now (bugfix for CPU and GPU) and the tests are passing.

**Comment by charleskawczynski** [comment]
Ok, fixing the test for VIJFH is trivial and only in the test suite. It does seem that we broke broadcasting for IJF, which is still used in the dss. A simple fix for now would be to dispatch knl_copyto and fix that part later. I think it’s just a getibdex issue, I’ll extend the test suite and see if we can also fix this today

**Comment by simonbyrne** [comment]
bors try

**Pull request merged**